### ISSUE TITLE
WGSL: Support the `unrestricted_pointer_parameters` language extension

### ISSUE THREAD

**Issue by guusw** [gfx-rs/wgpu#5158]
I tried searching a bit but could find out which is correct/outdated, some previous comments seem to mention that pointers to storage are illegal however I couldn't find anything about it in the spec (if so, please point me to it, thanks)

From the spec https://www.w3.org/TR/WGSL/#example-eab7f633

```wgsl
fn baz2(p : ptr<storage, f32>) {
}
```

which seems to not be accepted by naga:

```
Function [1] 'baz2' is invalid:
        Argument 'p' at index 0 is a pointer of space Storage { access: StorageAccess(LOAD) }, which can't be passed into functions.
```

**Comment by teoxoy** [comment]
We haven't yet implemented the `unrestricted_pointer_parameters` WGSL language extension.

**Comment by Alphapage** [comment]
Any news about this crucial feature ?
When do you plan to release `the unrestricted_pointer_parameters` WGSL language extension ?
Is there a way to bypass the limitation ?

**Comment by ErichDonGubler** [comment]
@Alphapage: Can you elaborate on why this functionality is "crucial"? It's technically an optional feature in the spec., and we on the Firefox team are trying to determine how to prioritize it.

**Comment by ErichDonGubler** [comment]
Firefox bug tracking this: [bug 1952974](https://bugzilla.mozilla.org/show_bug.cgi?id=1952974)

**Comment by fazil47** [comment]
@ErichDonGubler I was trying to use [WESL imports](https://wesl-lang.dev/spec/Imports) to reduce duplication, but any function in another file that requires access to a storage buffer can't do that without `unrestricted_pointer_parameters`.

**Comment by rnikander** [comment]
I ran into this, and the message from the compiler:
```
= note: Let Naga maintainers know that you ran into this at <https://github.com/gfx-rs/wgpu/issues/5158>, so they can prioritize it!
```
I'm learning WGPU/WGSL as I port some Swift+Metal stuff to Rust+WGPU. I'm no shader guru for sure, so maybe I can find a better way to do things, but right now if I just try to translate my old Metal shader code it would be nice to have this. I'm working on a drawing/painting app, and I had structs passed in uniform buffers that described gradients and brush color data. So helper functions to compute vertex or fragment color were (in Metal) taking references/pointers to these structs.

**Comment by rnikander** [comment]
Maybe there is no point, if I can pass structs by "value" but after optimization they are not really copied, if they are constants from uniforms. (?)

**Comment by lokegustafsson** [comment]
I would find this feature useful for defining helper functions when dealing with atomics. Concretely, I am doing subgroup-aggregrated atomics along the lines of
```wgsl
fn subgroupAtomicAdd(atomic_ptr: ptr<storage, atomic<u32>, read_write>, increment: u32) -> u32 {
  var ret: u32;
  if (subgroupElect()) {
    ret = atomicAdd(atomic_ptr, increment);
  }
  return subgroupBroadcastFirst(ret);
}
```
(since subgroupElect is not currently supported, I am also defining it manually)

**Comment by hukasu** [comment]
I hit that just today, thumbing up to show support

**Comment by gszep** [comment]
I ran into this today in an attempt to migrate from JS/WebGPU -> Rust/WGPU.

Without this language feature there would be a lot more code duplication. This feature is enabled by default when using JS/WebGPU with Chrome but runs into a compiler error with Rust/WGPU. I would love to be able to define functions like
```wgsl
fn constraint_update(idx: u32, array_of_structs: ptr<storage, array<MyStruct>, read_write>) {
  let this_struct = array_of_structs[idx];
  ...

  array_of_structs[idx] = that_struct;
}
```
This, together with `#import` statements for including other WGSL files would enable modular development and re-use of complex compute shaders <3

**Comment by TheBlek** [comment]
Would there be any performance penalty when using this extension?

Asking this because its marked as XXL feature, so a lot has to be done. I presume that's because storage pointers break some assumptions that can be made with `function` and `private` pointers. (Probably aliasing?) Would that mean the compiler will have problems optimizing storage pointers in some cases? Or it all can be mitigated but it will take a lot of work?

Sorry for dumb questions, I'm just curious.

**Comment by inner-daemons** [comment]
@TheBlek If an implementation would result in regressions to existing code, we would be wary varying of merging it. But its a language extension, which means its meant to be supported everywhere with very little penalty (and is just something browsers need to get around to implementing).

Are you interested in working on this or just curious about what it would bring?

**Comment by TheBlek** [comment]
@inner-daemons Well, I'm new to the project and the feature is marked as big so I figured it's not really my thing to do. As you can see by the questions I'm not even sure why this feature is so big 😅 

However if someone is willing to point in the right direction, I can give it a try. At least assess if I'll handle the complexity.

**Comment by inner-daemons** [comment]
@TheBlek Honestly this probably requires deep knowledge of WGSL and GPU memory models, so I wouldn't recommend trying it unless you are willing to spend a lot of time.

**Comment by haraldwer** [comment]
HarfBuzz-GPU relies on this feature for their SLUG implementation, just a heads up!
[This is their webgpu demo where this feature is used.](https://github.com/harfbuzz/harfbuzz/blob/0c962165/util/gpu/web/hb-gpu-web-webgpu.cc)

**Comment by inner-daemons** [comment]
For anyone curious, this would require a lot more complexity in naga. Currently our shader compiler is single-pass, meaning it doesn't do any transformations like optimizations or inlining of functions. This makes it quite likely the fastest shader compiler out there, by a wide margin.

However, enabling this unrestricted_pointer_parameters extension would require an inlining pass (which makes it so that the output only has one function). This is because the underlying languages (SPIR-V, etc) don't always support passing pointers around between functions, so we have to keep everything in just one function, the entry point. Adding an inlining pass would 1. make the compiler slower, 2. make the output not correspond as directly with the input and 3. increase complexity by quite a bit.

To understand this complexity increase, you have to understand how the underlying languages work. SPIR-V is bascially always inlined by the driver, so the driver needs very good information about all of the paths the program can take. This makes goto statements, for example, impossible, and spawning edge cases like the following:
```wgsl
fn a(input: i32) -> i32 {
    for (var i = 0; i < 10; i++) {
        for (var j = 0; j < 10; j++) {
            if i * j == input {
                return i;
            }
        }
    }
    return 0;
}
```
The point here is that in SPIR-V, with `break` statements, you can only break out of the innermost for loop, but this restriction does not apply to `return` statements. You would therefore have to inline it into something like the following:
```wgsl
var output = 0;
for (var i = 0; i < 10; i++) {
    var needs_exit = false;
    for (var j = 0; j < 10; j++) {
        if i * j == input {
            output = i;
            needs_exit = true;
            break;
        }
    }
    if needs_exit {
        break;
    }
}
```

The current plan is to at some point implement this inlining pass. Once wgpu is mostly spec compliant the unrestricted_pointer_parameters extension is a somewhat high priority. But implementing it would take a lot of time to get right and could slow down compiling by having to add another pass to completely rewrite the shader.

**All of the above is my understanding and Mozilla might have different plans than I expressed here

**Comment by gszep** [comment]
Thank you for the detailed response it's good to understand internals but explained in a simple way :)

From my side agentic coding has made this a non-issue. In fact I'm leaning towards saying don't implement this extension in favour of fast hot-reloading to keep the code generation - verification loop As tight as possible.

As long as agentic coding tools can easily write WebGPU, navigate the docs etc we're golden. I hope Mozilla sees this as a new priority - e.g. create an agent-friendly version of the docs etc.

**Comment by inner-daemons** [comment]
I want to experiment with solving this myself soon. Even if I am successful it will no doubt be a long time before it is enabled in Firefox. Just leaving it here so people know the status.

**Comment by schell** [comment]
o/ Hi all!

I'm working on a grant project which includes working with slabs, a lot. It would be great ergonomically if I could pass storage pointers around, since they are the slabs.
 
Following up on the discussion about the inlining pass approach.

@inner-daemons - I've been thinking about whether full inlining is the only option, or if there's an approach that doesn't introduce the `return`/`break` complexity, and I realized that it might be a similar problem I've encountered with adding generics to my `wgsl-rs` crate and adding atomics to the SPIR-V backend. It seems the core problem is that HLSL can't express `ptr<storage, T>` as a function parameter, but SPIR-V, MSL, and WGSL (with the ext) can. So the inlining pass is really an HLSL (and possibly edge-case SPIR-V) workaround being applied to all backends.

What if instead of inlining all functions into the entry point, we monomorphize only the functions that have storage pointer parameters?

So for each function with `ptr<storage, T>` parameters, treat the storage binding as a "template" parameter (or a "generic" parameter). At each call site passing a specific buffer, stamp out a concrete function where the buffer is referenced directly by name (not passed as a pointer). The other parameters like `offset` or whatever stay as regular parameters. Rewrite the call site to use the mangled name with the buffer arg dropped.

Example:
```wgsl
fn slab_read(ptr: ptr<storage, u32>, offset: u32) -> u32 {
    return (*ptr)[offset];
}
fn main() {
    let a = slab_read(&SLAB, 0);
    let b = slab_read(&OTHER, 4);
}
```
And after monomorphization:
```rust
fn slab_read_SLAB(offset: u32) -> u32 {
    return SLAB[offset];
}
fn slab_read_OTHER(offset: u32) -> u32 {
    return OTHER[offset];
}
fn main() {
    let a = slab_read_SLAB(0);
    let b = slab_read_OTHER(4);
}
```

This eliminates pointer parameters entirely and I think all backends can emit this. HLSL handles it fine because the buffer is referenced by name inside the function body, not passed as a pointer.

- Control flow is unchanged: return inside nested loops stays as return. No flag + break cascade needed.
- Call graph is preserved: there are more functions, not fewer. Output is still modular.
- Dedup: call sites passing the same buffer share one specialization.
- Transitive: if a specialized function calls another function with a storage pointer, cascade the specialization (same as generic type monomorphization).
- No architecture change: this is a post-validation transformation pass, analogous to the atomic type upgrade pass I wrote in `naga/src/front/atomic_upgrade.rs` (#5775). Record which functions need specialization during validation, transform post-validation.

It might end up being a smaller change, and easier to reason about?

There's also a natural path to delivering the feature in phases:
1. WGSL->WGSL only: just relax the validator to accept `ptr<storage, T>` params when the extension is enabled. No transformation needed! The WGSL backend emits pointer types as-is. Works for WebGPU + CPU validation immediately.
2. Monomorphization for non-WGSL backends: add the specialization pass for SPIR-V/HLSL/MSL. Enables native execution.

I'd be interested in working on this if there's appetite. I might be able to roll it into my grant project. The WGSL->WGSL path (phase 1) could be a quick win to unblock downstream tools that only need CPU-side validation and web deployment.

**Comment by inner-daemons** [comment]
@schell I think that an inlining pass would be the best. It would be most clear what is happening to downstream compilers and would minimize shader size. I started working on such a pass and I'm convinced it is possible. Yours is not a bad idea but I'm not sure it would be beneficial enough.

**Comment by schell** [comment]
Just to clarify a couple points from my idea:

Re "most clear what is happening to downstream compilers": I agree inlined output is easier for the driver, but since SPIR-V drivers inline anyway, I'd expect the driver to handle function calls fine. `naga`'s job is producing valid target code, not pre-optimizing (forgive me if I'm misspeaking for the maintainers, lol). The driver's optimizer is better positioned to make inlining decisions.

Re "minimize shader size": I think it might be the other way around? If `slab_read` is called from 20 sites with 2 different buffers, inlining duplicates the body 20 times inside main, while monomorphization produces 2 small functions called from 20 sites. But I could be missing something about the cases you're considering.

Either way, I'm not suggesting monomorphization and your inlining work are mutually exclusive. My idea could be used as a stepping stone to ship sooner. Phase 1 (WGSL->WGSL validator relaxation) is a tiny change that unblocks downstream tools needing CPU-side validation today. Phase 2 (monomorphization for non-WGSL backends) could ship while the inlining pass matures. If the inlining pass turns out to be cleaner in the long run, monomorphization could be removed, no harm done. 🤷 

@inner-daemons, can you elaborate on what you mean by 
> I'm not sure it would be beneficial enough

I might be missing something important :)

**Comment by inner-daemons** [comment]
Give me more time to think about it. I would assume that an inlining pass would be simpler and easier to write. I also remember another use case that the extension allows that might not be covered by your thing? Unsure.

You're probably right about the shader size, not sure what I was thinking there.

If you want to prove me wrong about anything go ahead.

**Comment by inner-daemons** [comment]
Alright @schell I think that your idea is exactly what @jimblandy had planned, I talked with him briefly and we disagreed but didn't come to a conclusion. I don't see any reason why yours wouldn't work but I suspect it could be messy.

The other restriction lifted by the extension would also be fixed by this I think, and I guess in theory you might have to copy-paste the function fewer times.

One other reason I lean toward inlining is that tint already does this and it works well for them. I feel that inlining would give us more leeway in the future to use HLSL how we want. I worked on mesh shaders in particular and the way that task payload variables had to be passed between functions was sorta disgusting (feel free to read that code if you want). I went back and forth with cwfitzgerald a bunch and decided this was the cleanest still-gross way to handle it. I suspect already-inlined functions would've massively helped there, and in general we might be able to rewrite the backends in a cleaner way in the future if we don't worry about function calls.

If you really do think your method is the simpler way for now, and you are willing to put up a PR, I'd be happy to review it and help push it through.

**Comment by schell** [comment]
@inner-daemons - great! Thank you 🙇 . I'll start by talking to my grant manager about it, then get a PR up 👍 .

**Comment by inner-daemons** [comment]
I will bring this up at our next maintainers' meeting since I don't speak for everybody. It will happen on Wednesday so it may be best for you to wait until then. Up to you obv.

**Comment by schell** [comment]
Thanks @inner-daemons - I'll wait as I am also waiting for my grant manager to get back to me 👍 .

**Comment by schell** [comment]
It looks like I will be able to help tackle this, @inner-daemons . Feel free to add me to the assignees (if you like) and I look forward to hearing from you about the maintainers' meeting 🙇 .

**Comment by inner-daemons** [comment]
@schell Great to hear! The maintainers' meeting is public by the way if you are interested in attending yourself (see [CONTRIBUTING.md](https://github.com/gfx-rs/wgpu/blob/trunk/CONTRIBUTING.md).

**Comment by schell** [comment]
I would love to, but I'm in New Zealand and that would be a very hard time to make. All good, though. I'll wait to hear from you and the team 🙇

**Comment by inner-daemons** [comment]
For those following along, Jim and Schell will discuss this in private. There were some concerns about a grant being given for what is more of a research project than an actual patch at this point. This extension would be very important to Mozilla and they would like to avoid bugs so they may wish for their own employees or someone closer to the project to tackle it. As always the standard for webgpu functionality exposed by Firefox is much higher than for native-only extensions due to security concerns. They may also decide to go with the inlining rather than this monomorphizing approach. Everything is up in the air but it is at least being discussed.

**Comment by jimblandy** [comment]
> I'm in New Zealand and that would be a very hard time to make.

We should probably have some wgpu maintainers' meetings timed so that it's practical for people in the Asia-Pacific time zones to attend. It would be great to be able to talk about this live.

> discuss this in private

@inner-daemons is accurately conveying what I said, but actually there's no reason the discussion should be private; it's just that the conversation began in chat. Having it here is better. I'll paraphrase our discussion below.

As to the larger plan here:

Supporting `unrestricted_pointer_parameters` is important for keeping Firefox's WebGPU effectively compatible with Chrome's. In principle, it's an option, so web developers should ideally write their code to work with or without it. In practice, if it's available in Chrome, that's most of the web developers' audience, so they may not bother. So while it's optional in theory, it's required in practice.

My concern is that I think Naga's present architecture cannot accommodate this feature without adding quite a lot of complexity to the back ends, which are already too complicated for most people to work on. Taking on technical debt is fine if it's done strategically, but my sense is that this would put us in dangerous territory.

The architectural change that we've been talking about is introducing a "backend-oriented IR". The idea is that Naga should have a second IR, generated after validation, that closely matches the backend languages (perhaps a sort of loose union of them, with unification of language features where appropriate), such that all backends can be simplified to just be straightforward text generators. Then, there should be a lowering pass from `naga::ir::Module` to this backend-oriented IR. The lowering pass would be shared by all the backends; each backend would pass the options saying which lowerings it needs.

I don't seem to have an issue filed for this. I've put filing one on my list.

**Comment by jimblandy** [comment]
I think we are also lacking in exploration. Even though I'm not keen (that phrasing accurately conveys how sure I am) on having Naga do inlining, I am still interested in seeing @inner-daemons ' inliner because it gives us a hint as to what this kind of code would look like, and that can give us a better sense of our options.
